test_app/tests/authentication/utils/test_claims.py

```
@@ -2865,3 +2868,278 @@ def test_create_claims_three_maps_last_writer_wins(
     res = claims.create_claims(authenticator, 'username', {}, [])
 
     assert res['access_allowed'] is True, 'The last allow map (order=3, always) must win over the deny at order=2'
+
+
+# ---------------------------------------------------------------------------
+# Performance and correctness tests for optimized claims processing
+# ---------------------------------------------------------------------------
+
+# Deterministically-generated group names for performance tests.
+# 33 groups matches the scale observed in production SAML responses.
+_MANY_GROUPS = [f"group-{hashlib.md5(f'seed-{i}'.encode()).hexdigest()[:16]}" for i in range(33)]
```

Ruff (0.15.18)

[error] Probable use of insecure hash functions in `hashlib`: `md5`

(S324)